### PR TITLE
Add pylint and mypy for gitlab-ci 

### PR DESCRIPTION
**Pull request by teoti001**
Pylint and mypy have been added to run in the CI-system during the test stage. Some minor fixes were necessary to make the code pass the tests without errors.  

Based on #71

### CODE REVIEW

File: caramel/models.py
Comment (Spindel) [inline caramel/models.py:3]: Why is this one here?  file-wide disables should usually have a reasoning, and the commit message conflicts two different changes at once, making blame view less helpful.  

ie. the change is valid, but the documentation of it is less so, as having it in one file may cause others to just copy-paste it as a "valid fix" when there isn't a clear reason for why it's disabled.

**Review by Spindel — CHANGES_REQUESTED** [review]
I think the mypy config change could just as well be placed in setup.cfg as that matches how other tools have been configured (flake8, nosetests, etc) and also keeps the changes down in the main part.

File: mypy.ini
Comment (teoti001) [reply mypy.ini:10]: Correct, missing type hints( and no stubs). From my understanding this is how you are supposed to handle pyramid/zope.transaction/sqlachemy since they dont have type hints or stubs for mypy.

File: caramel/models.py
Comment (teoti001) [reply caramel/models.py:3]: I'll move this to setup.cfg and restrict it to the effected areas.

File: .gitlab-ci.yml
Comment (teoti001) [reply .gitlab-ci.yml]: I compleatly missed that, thought that it was unofficial since mypy didn't find it on its own. Anyway now we have a fun error with mypy where it crashes on an internal error. 

``` caramel/models.py:246: error: INTERNAL ERROR -- Please try using mypy master on Github:
https://mypy.readthedocs.io/en/stable/common_issues.html#using-a-development-mypy-build
Please report a bug at https://github.com/python/mypy/issues
version: 0.910
Traceback (most recent call last):
  File "mypy/semanal.py", line 4872, in accept
  File "mypy/nodes.py", line 950, in accept
  File "mypy/semanal.py", line 1056, in visit_class_def
  File "mypy/semanal.py", line 1134, in analyze_class
  File "mypy/semanal.py", line 1143, in analyze_class_body_common
  File "mypy/semanal.py", line 1203, in apply_class_plugin_hooks
  File "/usr/local/lib/python3.8/dist-packages/SQLAlchemy-1.4.3-py3.8-linux-x86_64.egg/sqlalchemy/ext/mypy/plugin.py", line 173, in _base_cls_hook
    decl_class._scan_declarative_assignments_and_apply_types(ctx.cls, ctx.api)
  File "/usr/local/lib/python3.8/dist-packages/SQLAlchemy-1.4.3-py3.8-linux-x86_64.egg/sqlalchemy/ext/mypy/decl_class.py", line 123, in _scan_declarative_assignments_and_apply_types
    _scan_declarative_assignment_stmt(cls, api, stmt, cls_metadata)
  File "/usr/local/lib/python3.8/dist-packages/SQLAlchemy-1.4.3-py3.8-linux-x86_64.egg/sqlalchemy/ext/mypy/decl_class.py", line 377, in _scan_declarative_assignment_stmt
    type_id = names._type_id_for_callee(stmt.rvalue.callee)
  File "/usr/local/lib/python3.8/dist-packages/SQLAlchemy-1.4.3-py3.8-linux-x86_64.egg/sqlalchemy/ext/mypy/names.py", line 168, in _type_id_for_callee
    return _type_id_for_funcdef(callee.node)
  File "/usr/local/lib/python3.8/dist-packages/SQLAlchemy-1.4.3-py3.8-linux-x86_64.egg/sqlalchemy/ext/mypy/names.py", line 179, in _type_id_for_funcdef
    if hasattr(node.type.ret_type, "type"):
AttributeError: 'NoneType' object has no attribute 'ret_type'
caramel/models.py:246: : note: use --pdb to drop into pdb ```

I don't really understand it, something to do with the models.AccessLog class( the Columns ?)

Comment (teoti001) [reply .gitlab-ci.yml]: Outdated sqlalchemy install on my end(1.4.3), version 1.4.20 works.

File: mypy.ini
Comment (teoti001) [reply mypy.ini]: > I think this is missing sqlalchemy[mypy] to get the type hints properly. ( As I don't think mypy --install-types adds that feature)
> 
> Details on sqlalchemy docs here:
> 
> https://docs.sqlalchemy.org/en/14/orm/extensions/mypy.html

sqlachemy does in fact have a mypy stub, removed the ignore for sqlalchemy.

File: setup.cfg
Comment (Spindel) [inline setup.cfg:48]: No newline at EOF makes future diff sad.

**Comment by Spindel** [comment]
LGTM except the newline, I could fix that up or you, your call.

**Comment by Spindel** [comment]
> LGTM except the newline, I could fix that up or you, your call.


Okay. the LGTM is assuming as I did ignore the commits that were shared between the other branch and this, so either way this one will have some conflicts when/if those changes are done.   Sorry for being unclear there.

File: setup.cfg
Comment (teoti001) [reply setup.cfg]: fixed

File: .gitlab-ci.yml
Comment (Spindel) [inline .gitlab-ci.yml:70]: This one still seems wrong.  sqlalchemy2-stubs is from  dropbox and is not the same stubs that the official sqlalchemy has, including the plugin. It's also mostly unmaintained ( I know from experience ).

It should be `SQLAlchemy[mypy] >= 1.4`   or `~= 1.4` instead.

File: setup.cfg
Comment (Spindel) [inline setup.cfg:35]: The reason I'm commenting above, is that down here it's _not_ activating the plugin that was installed from  `sqlalchemy2-stubs` but rather using the one in `SQLAlchemy[mypy]`.

File: caramel/models.py
Comment (Spindel) [inline caramel/models.py:7]: I think this will cause issues in the future with pylint or flake8.
Better to formulate it  as:

from  from sqlalchemy.ext.declarative import  ( declared_attr,  as_declarative) 

And just remove the _ from declared_attr below.

File: caramel/models.py
Comment (Spindel) [inline caramel/models.py:135]: This doesn't really match the type expected, it should be:

from typing import List


    accessed: List[AccessLog]  = _orm.relationship(
        "AccessLog", backref="csr", order_by="AccessLog.when.desc()"
    )
    certificates: List[Certificate] = _orm.relationship(
        "Certificate",
        backref="csr",
        order_by="Certificate.not_after.desc()",
        lazy="dynamic",
        cascade="all, delete-orphan",
    )

File: setup.cfg
Comment (Spindel) [inline setup.cfg:48]: this last one, `cls`,  looks a bit dubious to me as it's commonly used on classmethods to reference the class rather than the instance (cls vs self)  so it's name showing up here gives me a bit of a red flag.

File: caramel/models.py
Comment (Spindel) [reply caramel/models.py:7]: Also. This change ("sqlalchemy.ext.declarative") is not matched in setup.py below, the versions don't match, and SQLAlchemy 1.1 will fail here.  So the dep in setup.py should be bumped to `>= 1.4`

**Review by Spindel — CHANGES_REQUESTED** [review]
It needs a rebase on master as it's parent is merged,  ( damn you github UI for not showing me the diff against master branch, but rather something else.  )

Other than that, a few type related changes should be done.

File: .gitlab-ci.yml
Comment (teoti001) [reply .gitlab-ci.yml:70]: Reason for this is that pip install sqlalchemy[mypy] just gave me errors saying it couldn't find it, turns out you have to put it in quotes( "sqlachemy[mypy]"). Fixed

File: setup.cfg
Comment (teoti001) [reply setup.cfg:35]: no more sqlalchemy2-stub, solved

File: caramel/models.py
Comment (teoti001) [reply caramel/models.py:7]: SQLalchemys changelog mentions sqlalchemy.ext.declarative in an example snippet in the [1.1 changelog](https://docs.sqlalchemy.org/en/13/changelog/migration_11.html) so I don't think this should be a problem.

File: caramel/models.py
Comment (teoti001) [reply caramel/models.py:135]: The problem here is that Certificate and AccessLog is used before assignment, which pylint makes a fuss about, and changing the order to CSR-class after Certificate and AccessLog we have the same problem, because AccessLog uses CSR. One option is to say they are lists of Base but that sounded to prone for confusion("Base" is very generic as a name) so thats why it ended up like that. I guess I could suppress the error, but that feels naughty.

File: setup.cfg
Comment (teoti001) [reply setup.cfg:48]: The line 94(maybe 95) in models.py

 ```    
def __tablename__(cls): # pylint: disable=no-self-argument
        return cls.__name__.lower() 
```
triggers the error "E1101: Instance of 'Base' has no '__name__' member (no-member)", but from my understanding it gets generated when the decoration is done(cls turned from a reference to an instance(a "self") and to the class), Base is already in the list there but that covers to errors later down concerning the metadata member. So I could disable the error for the line, but in my head this was more logical, though not very.

File: caramel/models.py
Comment (Spindel) [reply caramel/models.py:135]: Then the fix is to do:

`List["Accesslog"]`    Quotes around the class name makes it a safe forward-declaration.

File: setup.cfg
Comment (Spindel) [reply setup.cfg:48]: For this case, as it's not a truth for "all" cls,  I'd say make it happen on the line.

The other ones ought to be true for all instances by that name, so I'd call those "ok" for this setting.

File: caramel/models.py
Comment (teoti001) [reply caramel/models.py:135]: Didn't know that, thanks! will do

**Review by Spindel — APPROVED** [review]
LGTM.

**Pull request merged**
